src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -0,0 +1,254 @@
+            Ok(workdir)
+        }
+    }
+
+    fn calculate_nailgun_name(main_class: &String, digest: &Digest) -> String {
+        format!("{}_{}", main_class, digest.0)
+    }
+}
+
+impl super::CommandRunner for NailgunCommandRunner {
+    fn run(
+        &self,
+        req: MultiPlatformExecuteProcessRequest,
+        context: Context) -> BoxFuture<FallibleExecuteProcessResult, String> {
+
+        let mut client_req = self.extract_compatible_request(&req).unwrap();
+
+        if !client_req.is_nailgunnable {
+            trace!("The request is not nailgunnable! Short-circuiting to regular process execution");
+            return self.inner.run(req, context)
+        }
+        debug!("Running request under nailgun:\n {:#?}", &client_req);
+
+        // Separate argument lists, to form distinct EPRs for (1) starting the nailgun server and (2) running the client in it.
+        let ParsedArgLists {nailgun_args, client_args } = split_args(&client_req.argv);
+        let nailgun_req = get_nailgun_request(nailgun_args, client_req.input_files, client_req.jdk_home.clone());
+        trace!("Extracted nailgun request:\n {:#?}", &nailgun_req);
+
+        let maybe_jdk_home = nailgun_req.jdk_home.clone();
+
+        let main_class = client_args.iter().next().unwrap().clone(); // We assume the last one is the main class name
+        let nailgun_req_digest = crate::digest(MultiPlatformExecuteProcessRequest::from(nailgun_req.clone()), &self.metadata);
+        let nailgun_name = NailgunCommandRunner::calculate_nailgun_name(&main_class, &nailgun_req_digest);
+        let nailgun_name2 = nailgun_name.clone();
+        let nailgun_name3 = nailgun_name.clone();
+
+        let nailguns_workdir = try_future!(self.get_nailguns_workdir(&nailgun_name));
+
+        // Materialize the directory for running the nailgun server, if we need to.
+        let workdir_path2 = nailguns_workdir.clone();
+        let workdir_path3= nailguns_workdir.clone();
+        let workdir_path4 = nailguns_workdir.clone();
+        let materialize = self.inner
+            .store
+            // TODO This materializes the input files in the client req, which is a superset of the files we need (we only need the classpath, not the input files)
+            .materialize_directory(nailguns_workdir.clone(), client_req.input_files, context.workunit_store.clone())
+            .and_then(move |_metadata| {
+                maybe_jdk_home.map_or(Ok(()), |jdk_home_relpath| {
+                    let jdk_home_in_workdir = workdir_path2.clone().join(".jdk");
+                    if !jdk_home_in_workdir.exists() {
+                        symlink(jdk_home_relpath, jdk_home_in_workdir)
+                            .map_err(|err| format!("Error making symlink for local execution in workdir {:?}: {:?}", &workdir_path2, err))
+                    } else {
+                        debug!("JDK home for Nailgun already exists in {:?}. Using that one.", &workdir_path4);
```

Comment:
We should probably assert that the JDK paths are the same, or restart the nailgun if they're not, rather than silently re-using maybe the wrong JDK...